test/parallel/test-timers-same-timeout-wrong-list-deleted.js

```diff
@@ -0,0 +1,62 @@
+'use strict';
+
+/*
+ * This is a regression test for https://github.com/nodejs/node/issues/7722.
+ *
+ * When nested timers have the same timeout, calling clearTimeout on the
+ * older timer after it has fired causes the list the newer timer is in
+ * to be deleted. Since the newer timer was not cleared, it still blocks
+ * the event loop completing for the duration of its timeout, however, since
+ * no reference exists to it in its list, it cannot be canceled and its
+ * callback is not called when the timeout elapses.
+ */
+
+const common = require('../common');
+const assert = require('assert');
+const Timer = process.binding('timer_wrap').Timer;
+
+const TIMEOUT = 100;
+const start = Timer.now();
+
+// This bug also prevents the erroneously dereferenced timer's callback
+// from being called, so we can't use it's execution or lack thereof
+// to assert that the bug is fixed.
+process.on('exit', function() {
+  const end = Timer.now();
+  assert.equal(end - start < TIMEOUT * 2, true,
+               'Elapsed time does not include second timer\'s timeout.');
+});
+
+const handle1 = setTimeout(common.mustCall(function() {
+  // Cause the old TIMEOUT list to be deleted
+  clearTimeout(handle1);
+
+  // Cause a new list with the same key (TIMEOUT) to be created for this timer
+  const handle2 = setTimeout(function() {
+    common.fail('Inner callback is not called');
+  }, TIMEOUT);
+
+  setTimeout(common.mustCall(function() {
+    // Attempt to cancel the second timer. Fix for this bug will keep the
+    // newer timer from being dereferenced by keeping its list from being
+    // erroneously deleted. If we are able to cancel the timer successfully,
+    // the bug is fixed.
+    clearTimeout(handle2);
+    setImmediate(common.mustCall(function() {
+      setImmediate(common.mustCall(function() {
+        const activeHandles = process._getActiveHandles();
+        activeHandles.forEach(common.mustCall(function(handle) {
```

Comment:
> The assertion that there is still one Timer remaining at this point fails, though, presumably because of the two calls to setImmediate? There are zero in the list of active handles.

Absolutely, my mistake, and sorry for the confusion. It's expected that no timers are left in the list of active handles at this point.

Basically the goal is to have some check that `Timer` instances are actually added to the list of active handles at some point, otherwise checking for their absence doesn't tell us much. We know _now_ that these `Timer` instances are added to the list of active handles, because we've been working on that test and we checked that ourselves, but future changes might change that and the test might pass just because `Timer` instances are never added to the list of active handles anymore.

What do you think of adding a check after the [10 ms timer is added](https://github.com/nodejs/node/pull/7827/files#diff-d95655ffcc984cf8c55e8bf0388673f1R58) that makes sure that there are at least two instances of `Timer` in the active handles list, one with a delay of `TIMEOUT` and the other with a delay of 10ms?